### PR TITLE
Handle viewpost deeplink in view model

### PR DESCRIPTION
**Pull request by planarvoid**
This PR introduces 2 changes:
- moves handling of the `wordpress://viewpost` links to the viewmodel. They're now handled by the `ReaderLinkHandler` since they are opened in the Reader post detail. In addition to this I've changed the behaviour a little bit. Previously if the blog ID and post ID were not present/parsed, the app just finished. I've changed the behaviour to open default reader instead. I think this is better.
- Adds check whether the user is logged in before opening a link. If the user is not logged in, the URI is cached and the login is triggered for a result. Once we get the result, we retrigger the cached URL. This is a very generic approach and works for all the use cases. 

To test:
- Make sure you're logged in with a blog with remote ID X and post with remote ID Y
- open a `wordpress://viewpost?blogId=X&postId=Y` deeplink
- Notice the post detail is opened in the reader post detail
- Go back
- Notice you go to the main activity instead of closing the app

To test:
- Make sure you're logged out
- open a `wordpress://viewpost?blogId=X&postId=Y` deeplink
- Notice you're redirected to Login process
- Log in with a blog with remote ID X and post with remote ID Y
- Finish login
- Notice the post detail is opened in the reader post detail
- Go back
- Notice you go to the main activity instead of closing the app

To test:
- Make sure you're logged in
- open a `wordpress://viewpost` deeplink
- Notice the reader is opened

To test
- Test other links (stats, pages, read) without the user being logged in
- Notice the user is redirected to the desired link once login ends
- Notice the links work as expected when the user is logged in

## Regression Notes
1. Potential unintended areas of impact
- all the deeplinks

2. What I did to test those areas of impact (or what existing automated tests I relied on)
- I've tested multiple flows and everything is unit tested

3. What automated tests I added (or what prevented me from doing so)
- unit tests

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by renanferrari — APPROVED** [review]
Thanks for the changes! Code is looking good and even though I couldn't make everything work as expected, it seems that the odd behaviors I found already existed before, so feel free to merge this once you read the comments below.

>- Make sure you're logged in with a blog with remote ID X and post with remote ID Y
>- open a wordpress://viewpost?blogId=X&postId=Y deeplink

I don't know if I understood this correctly, but is this kind of link supposed to work only for posts from the user's sites? Or is it supposed to work for any posts? 

I assume it's the latter, but I couldn't make it work. I tested with posts from the user's sites and also with posts from other sites, but in both cases I kept getting redirected to the Reader's main screen.

Apparently this was not working before as well, the difference is that I'd only see the Toast instead of getting redirected, so feel free to open an issue to track this separately from this PR.

>- Make sure you're logged out
>- open a wordpress://viewpost?blogId=X&postId=Y deeplink

I couldn't test this for the `viewpost` links specifically (because of the issue I described above), but it did work well for other links I tested. (And I like that the solution is quite simple!) 

One thing I wonder if it wouldn't be possible to open the Reader while logged out. It looks like we already do this for other post links (like the ones in #14471). Since this would change the existing behavior, I'd also open an issue to track it separately from this PR.

**Comment by planarvoid** [comment]
thanks for the review @renanferrari !

> I don't know if I understood this correctly, but is this kind of link supposed to work only for posts from the user's sites? Or is it supposed to work for any posts?

> I assume it's the latter, but I couldn't make it work. I tested with posts from the user's sites and also with posts from other sites, but in both cases I kept getting redirected to the Reader's main screen.

> Apparently this was not working before as well, the difference is that I'd only see the Toast instead of getting redirected, so feel free to open an issue to track this separately from this PR.

I've just checked and you don't have to be logged in by the site X. The thing is that the post has to be published and the site has to be public. You can try the following deep link to a public post: `wordpress://viewpost?blogId=165726002&postId=1328` which should work for any account.

> One thing I wonder if it wouldn't be possible to open the Reader while logged out. It looks like we already do this for other post links (like the ones in #14471). Since this would change the existing behavior, I'd also open an issue to track it separately from this PR.

I think it should be possible, I've created an issue to handle this here https://github.com/wordpress-mobile/WordPress-Android/issues/14562

**Comment by renanferrari** [comment]
>I think it should be possible, I've created an issue to handle this here #14562

Thank you for creating the issue!

>The thing is that the post has to be published and the site has to be public.

I had tried that but it wasn't working, in the end I figured out that I had to escape the `&` character in the URL for the adb command to work, like so:

```
adb shell am start \
 -W -a android.intent.action.VIEW \
 -c android.intent.category.BROWSABLE \
 -d "wordpress://viewpost?blogId=165726002\&postId=1328"
```
 
Such a silly thing 🤦‍♂️

Everything is working as expected now! LGTM 👍

**Comment by planarvoid** [comment]
> I had tried that but it wasn't working, in the end I figured out that I had to escape the & character in the URL for the adb command to work, like so:

ah yeah, sorry I haven't mentioned that. It blocked me for a while as well :D

**Pull request merged**
